### PR TITLE
feat: redirect to previous path if its a 404

### PR DESCRIPTION
**Pull request by crowlKats**
This is an MVP, not the perfect result, but much better than what we currently have (which is nothing)

### CODE REVIEW

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -23,7 +23,7 @@ jobs:
       - name: Set up Deno
         uses: denoland/setup-deno@v2
         with:
-          deno-version: canary
+          deno-version: 2.0.6
```
Comment (crowlKats) [inline .github/workflows/deploy.yml:26]: done to prevent the upcoming deno doc changes to affect the reference API

**Comment by philhawksworth** [comment]
I like it!

In future I think we can have some sort of combination of this logic, and a suitably populated 404 page with onward links to likely locations the user intended... but for now, this would be a valuable improvement

**Comment by crowlKats** [comment]
our current 404 page is terrible (just an empty page that says "Not Found"), and since our URLs are structured hierarchically, going to the next possible parent makes the most sense.
One example also is for some reason google indexed https://docs.deno.com/api/node/streams, which doesnt even exist, so going to the parent in this case is a better experience

**Comment by josh-collinsworth** [comment]
I still think overriding the "correct" response is hazardous and unexpected behavior. Even if it works sometimes, it probably won't others, and that creates a confusing experience. (I also don't think it's probably a good thing that we make it impossible to even hit a 404, for a few reasons.)

I also notice that if the first redirect doesn't hit an existing page, it creates a "too many redirects" error.

Can we write some kind of "is this what you were looking for?" logic into the 404 page that pattern matches, or something along those lines? That seems much more in line with expected behavior. I just think this seems like a pretty haphazard solution.

**Comment by josh-collinsworth** [comment]
Happy to take a stab at this, if it helps

**Comment by BlackAsLight** [comment]
I think it may lead users to be confused about what just happened if there isn't something on the page telling them they were redirected due to a 404. 

Like if I clicked a link and it just loaded the exact same page I'm on, I'd think this page is broken and not that the page I'm trying to go to doesn't exist. It would feel like a buggy experience.

**Comment by crowlKats** [comment]
sure @josh-collinsworth, go ahead; i probably wont have have time to work on this anytime soon

File: .github/workflows/deploy.yml [hunk 2]
```diff
@@ -23,7 +23,7 @@ jobs:
       - name: Set up Deno
         uses: denoland/setup-deno@v2
         with:
-          deno-version: canary
+          deno-version: 2.0.5
```
Comment (marvinhagemeister) [inline .github/workflows/deploy.yml:26]: Nit: Seems like an unrelated change

**Review by marvinhagemeister — APPROVED** [review]
